### PR TITLE
Blog post on history of Service Weaver's core API.

### PR DESCRIPTION
**Pull request by mwhittaker**

### CODE REVIEW

File: website/blog/history.md [hunk 1]
```diff
@@ -0,0 +1,466 @@
...
+of Service Weaver over two years ago.  We hope that by studying the API's
+history, you gain a better understanding of
+
+- why the API is the way it is,
+- how it differs from alternative APIs, and
+- what the API might look like in the future.
+
+We conclude with some lessons learned on API design.
+
+## September 30, 2021: In the Beginning...
```
Comment (ghemawat) [inline website/blog/history.md:17]: Could benefit from an brief listing of all changes at the top (one line per change). Skimmers can just go with that.

Maybe table of conntents will show up automatically and provide that listing?

File: website/blog/history.md [hunk 2]
```diff
@@ -0,0 +1,466 @@
...
+    // ...
+}
+```
+
+This version of `weaver.Run` clearly runs the provided function (e.g., `serve`),
+and `weavertest.Local.Test` clearly does not.
+
+## Lessons Learned
+
+This brings us to the Service Weaver API we have today.
```
Comment (ghemawat) [inline website/blog/history.md:456]: Can you add in some ideas in rough form (they can be fleshed out in a subsequent change):

* Using an API shows potential problems, so use it before committing to it
* API design is hard: full of pros and cons that need to be balanced. In particular, fine line between boilerplate and asking users to be precise.
* New mini-languages (//weaver:... comments) are simple to start with, but easy to get into something complicated
* Declaring things ahead of time can be powerful (e.g., the component graph)

**Review by ghemawat — APPROVED** [review]
Looks great! Thanks for writing this up.

File: website/blog/history.md (hunk 2 above)
Comment (mwhittaker) [reply website/blog/history.md]: Added the "Lessons Learned" section you wrote.

File: website/blog/history.md (hunk 1 above)
Comment (mwhittaker) [reply website/blog/history.md]: Good idea! Done.

File: website/blog/history.md [hunk 3]
```diff
@@ -453,9 +453,32 @@ and `weavertest.Local.Test` clearly does not.
...
-This brings us to the Service Weaver API we have today.
-
-TODO: Summarize some lessons learned.
+This brings us to the Service Weaver API we have today. The most important
+lesson here is that good interface design is hard. We went back and forth on
+how to declare components. One thing that can often help is to model things
+after existing APIs. E.g., it is easy to define an API for a familiar concept
+like an associative container. However the problem we faced, embedding remote
+communication interfaces in the middle of Go, was unfamiliar and therefore we
+had several missteps. What really helped was taking the time to get experience
```
Comment (ghemawat) [inline website/blog/history.md:478]: s/had/made/

**Comment by mwhittaker** [comment]
Thanks Sanjay! We'll release this blog post in a couple of weeks, closer to Robert's GopherCon talk.

**Review by ghemawat — APPROVED** [review]
Great! Thanks for writing up most of this.

File: website/blog/history.md (hunk 3 above)
Comment (mwhittaker) [reply website/blog/history.md]: Done.

**Pull request merged**